core/src/main/java/org/infinispan/transaction/impl/TransactionCoordinator.java

```diff
@@ -99,92 +107,116 @@ public PrepareCommand createPrepareCommand(GlobalTransaction gtx, List<WriteComm
       }
    }
 
-   public final int prepare(LocalTransaction localTransaction) throws XAException {
+   public final CompletionStage<Integer> prepare(LocalTransaction localTransaction) {
       return prepare(localTransaction, false);
    }
 
-   public final int prepare(LocalTransaction localTransaction, boolean replayEntryWrapping) throws XAException {
-      validateNotMarkedForRollback(localTransaction);
+   public final CompletionStage<Integer> prepare(LocalTransaction localTransaction, boolean replayEntryWrapping) {
+      CompletionStage<Integer> markRollbackStage = validateNotMarkedForRollback(localTransaction);
+      if (markRollbackStage != null) {
+         return markRollbackStage;
+      }
 
       if (isOnePhaseCommit(localTransaction)) {
          if (trace) log.tracef("Received prepare for tx: %s. Skipping call as 1PC will be used.", localTransaction);
-         return XA_OK;
+         return XA_OKAY_STAGE;
       }
 
       PrepareCommand prepareCommand = commandCreator.createPrepareCommand(localTransaction.getGlobalTransaction(), localTransaction.getModifications(), false);
       if (trace) log.tracef("Sending prepare command through the chain: %s", prepareCommand);
 
       LocalTxInvocationContext ctx = icf.running().createTxInvocationContext(localTransaction);
       prepareCommand.setReplayEntryWrapping(replayEntryWrapping);
-      try {
-         invoker.running().invoke(ctx, prepareCommand);
+      CompletionStage<Object> prepareStage = invoker.running().invokeAsync(ctx, prepareCommand);
+      return CompletionStages.handleAndCompose(prepareStage, (ignore, prepareThrowable) -> {
+         if (prepareThrowable != null) {
+            if (shuttingDown)
+               log.trace("Exception while preparing back, probably because we're shutting down.");
+            else
+               log.errorProcessingPrepare(prepareThrowable);
+
+            //rollback transaction before throwing the exception as there's no guarantee the TM calls XAResource.rollback
+            //after prepare failed.
+            return CompletionStages.handleAndCompose(rollback(localTransaction), (ignore2, rollbackThrowable) -> {
```

Comment:
No need for `handleAndCompose`

Reply:
Not sure, why. We need to handle the case when it was not an error to wrap it with an XAException still. And we want to still catch the rollback exception to supress or rethrow that.

Reply:
Yes, but none of those need to return a CompletionStage, so you can use handle instead of handleAndCompose.

Reply:
I was trying to keep all the exceptions as bare `XAException`. If I do the other I would have to use `CompletionException` wrapping `XAException` and all callers would have to pay attention to that including `TransactionXAAdapter`, but I can do that.

Reply:
Actually I don't feel comfortable changing this right now. I can revisit later if we need.

Reply:
> I was trying to keep all the exceptions as bare `XAException`. If I do the other I would have to use `CompletionException` wrapping `XAException` and all callers would have to pay attention to that including `TransactionXAAdapter`, but I can do that.

Not sure what you mean. When you do `CompletableFuture.join()` it will wrap the exception in a `CompletionStage` anyway, so you have to be prepared to extract the exception with `CompletableFutures.extractException`.

But I'm ok with revisiting this later.